File: PARTAGER/ConvertirCoordonnees/ProjectionLambert_1.vb
Namespace Coordonnees
    ''' <summary> Ce module contient les fonctions permettant la conversion entre les coordonnées Géodétique (Latitude et Longitude) et une projection 
    ''' Lambert Conforme Conique à 1 parallèle en X et Y ou l'inverse traduit du projet Geotrans écrit en C </summary>
    Friend Module ProjectionLambert_1
#Region "Variables et Constantes"
        '''<summary> Donne par defaut les différentes constantes liées à la projection Lambert 1 et au système CLARKE 1880 ("NTF") </summary>
        Private Const MAX_LAT As Double = Deg_Rad * 89.9997222222222R ' ((PI * 89.9997222222222) / 180)
        Private Const MIN_SCALE_FACTOR As Double = 0.000000001R
        Private Const ONE_SECOND As Double = 0.00000489R
        Private Lambert_1_a As Double = RayonWGS84
        Private Lambert_1_f As Double = F_WGS84 ' 1 / 298.257223563
        Private es As Double = 0.0818191908426219R
        Private es_OVER_2 As Double = 0.040909595421311R
        Private Lambert_1_n As Double = 0.70710678118655R
        Private Lambert_1_rho0 As Double = 6388838.2901212R
        Private Lambert_1_rho_olat As Double = 6388838.2901211R
        Private Lambert_1_t0 As Double = 0.41618115138974R
        Private Lambert_1_Origin_Lat As Double = Deg_Rad * 45.0R ' (45 * PI / 180)
        Private Lambert_1_Origin_Long As Double = 0.0R
        Private Lambert_1_False_Northing As Double = 0.0R
        Private Lambert_1_False_Easting As Double = 0.0R
        Private Lambert_1_Scale_Factor As Double = 1.0R
        Private Const Lambert_Delta_Easting As Double = 400_000_000.0R
        Private Const Lambert_Delta_Northing As Double = 400_000_000.0R

        '''<summary> Donne différentes erreurs pouvant survenir lors de la transformation des coordonnées </summary>
        <Flags>
        Friend Enum LAMBERT_1 As Integer
            '''<summary>Flag OK</summary>"
            NO_ERROR = 0
            '''<summary>Flag erreur concernant la Latitude</summary>
            LAT_ERROR = 1
            '''<summary>Flag erreur concernant la Longitude</summary>
            LON_ERROR = 2
            '''<summary>Flag erreur concernant les abscisses(Easting ou X)</summary>
            EASTING_ERROR = 4
            '''<summary>Flag erreur concernant les ordonnées(Northing ou Y)</summary>
            NORTHING_ERROR = 8
            '''<summary>Flag erreur concernant l'origine de la Latitude</summary>
            ORIGIN_LAT_ERROR = 16
            '''<summary>Flag erreur concernant le méridien central</summary>
            CENT_MER_ERROR = 32
            '''<summary>Flag erreur concernant le facteur d'échelle</summary>
            SCALE_FACTOR_ERROR = 64
            '''<summary>Flag erreur concernant le demi axe de l'éllipsoïde</summary>
            A_ERROR = 128
            '''<summary>Flag erreur concernant l'applatissement de l'éllipsoïde</summary>
            INV_F_ERROR = 256
        End Enum
#End Region
#Region "Functions and Procedures"
        Private Function LAMBERT_m(clat As Double, essin As Double) As Double
            Return (clat / Math.Sqrt(1.0R - essin * essin))
        End Function
        Private Function LAMBERT_t(lat As Double, essin As Double) As Double
            Return Math.Tan(Pi_4 - lat / 2.0R) / Math.Pow((1.0R - essin) / (1.0R + essin), es_OVER_2)
        End Function
        Private Function ES_SIN_Func(sinlat As Double) As Double
            Return (es * sinlat)
        End Function
        '''<summary>The function Set_Lambert_1_Parameters receives the ellipsoid parameters and Lambert Conformal Conic (1 parallel)
        ''' projection parameters as inputs, and sets the corresponding state variables. </summary>
        '''<param name="a">Semi-major axis of ellipsoid, in meters(input)</param>
        '''<param name="f">Flattening of ellipsoid(input)</param>
        '''<param name="Origin_Latitude">Latitude of origin, in radians(input)</param>
        '''<param name="Central_Meridian">Longitude of origin, in radians(input)</param>
        '''<param name="False_Easting">False easting, in meters(input)</param>
        '''<param name="False_Northing">False northing, in meters(input)</param>
        '''<param name="Scale_Factor">Projection scale factor(input)</param>
        '''<returns>If any errors occur, the error code(s) are returned by the function, otherwise LAMBERT_1.NO_ERROR is returned.</returns>
        Friend Function Set_LAMBCC1_Parameters(a As Double, f As Double, Origin_Latitude As Double, Central_Meridian As Double,
                                                         False_Easting As Double, False_Northing As Double, Scale_Factor As Double) As LAMBERT_1
            Dim es2 As Double, es_sin As Double, m0 As Double, inv_f As Double = 1.0R / f
            Dim Error_Code As LAMBERT_1
            If a <= 0 Then
                Error_Code = Error_Code Or LAMBERT_1.A_ERROR
            End If
            If (inv_f < 250.0R) OrElse (inv_f > 350.0R) Then
                Error_Code = Error_Code Or LAMBERT_1.INV_F_ERROR
            End If
            If ((Origin_Latitude < -MAX_LAT) OrElse (Origin_Latitude > MAX_LAT)) OrElse (Origin_Latitude > -ONE_SECOND) AndAlso (Origin_Latitude < ONE_SECOND) Then
                Error_Code = Error_Code Or LAMBERT_1.ORIGIN_LAT_ERROR
            End If
            If (Central_Meridian < -PI) OrElse (Central_Meridian > TwoPI) Then
                Error_Code = Error_Code Or LAMBERT_1.CENT_MER_ERROR
            End If
            If Scale_Factor < MIN_SCALE_FACTOR Then
                Error_Code = Error_Code Or LAMBERT_1.SCALE_FACTOR_ERROR
            End If
            If Error_Code = LAMBERT_1.NO_ERROR Then
                Lambert_1_a = a
                Lambert_1_f = f
                Lambert_1_Origin_Lat = Origin_Latitude
                If Central_Meridian > PI Then
                    Central_Meridian -= TwoPI
                End If
                Lambert_1_Origin_Long = Central_Meridian
                Lambert_1_False_Easting = False_Easting
                Lambert_1_False_Northing = False_Northing
                Lambert_1_Scale_Factor = Scale_Factor
                es2 = 2.0R * Lambert_1_f - Lambert_1_f * Lambert_1_f
                es = Math.Sqrt(es2)
                es_OVER_2 = es / 2.0R
                Lambert_1_n = Math.Sin(Lambert_1_Origin_Lat)
                es_sin = ES_SIN_Func(Math.Sin(Lambert_1_Origin_Lat))
                m0 = LAMBERT_m(Math.Cos(Lambert_1_Origin_Lat), es_sin)
                Lambert_1_t0 = LAMBERT_t(Lambert_1_Origin_Lat, es_sin)
                Lambert_1_rho0 = Lambert_1_a * Lambert_1_Scale_Factor * m0 / Lambert_1_n
                Lambert_1_rho_olat = Lambert_1_rho0
            End If
            Return Error_Code
        End Function
        '''<summary>The function Convert_Lambert_1_To_Geodetic converts Lambert ConformalConic (1 parallel) projection (easting and northing) coordinates to Geodetic
        '''(latitude and longitude) coordinates, according to the current ellipsoid and Lambert Conformal Conic (1 parallel) projection parameters.</summary>
        '''<param name="PointGrille"> point à convertir </param>
        '''<param name="PointLatLon"> resultat de la convertion </param>
        '''<returns>If any errors occur, the error code(s) are returned by the function PointLatLon is empty, otherwise LAMBERT_NO_ERROR is returned.</returns>
        Friend Function Convert_LAMBCC1_To_Geodetic(PointGrille As PointD, ByRef PointLatLon As PointD) As LAMBERT_1
            Dim dx, dy, rho, rho_olat_MINUS_dy, t, PHI, es_sin, tempPHI, theta As Double
            Dim tolerance As Double = 0.000000000485R
            Dim count As Integer = 30
            Dim Error_Code As LAMBERT_1 = LAMBERT_1.NO_ERROR
            If (PointGrille.X < (Lambert_1_False_Easting - Lambert_Delta_Easting)) OrElse (PointGrille.X > (Lambert_1_False_Easting + Lambert_Delta_Easting)) Then
                Error_Code = Error_Code Or LAMBERT_1.EASTING_ERROR
            End If
            If (PointGrille.Y < (Lambert_1_False_Northing - Lambert_Delta_Northing)) OrElse (PointGrille.Y > (Lambert_1_False_Northing + Lambert_Delta_Northing)) Then
                Error_Code = Error_Code Or LAMBERT_1.NORTHING_ERROR
            End If
            If Error_Code = LAMBERT_1.NO_ERROR Then
                dy = PointGrille.Y - Lambert_1_False_Northing
                dx = PointGrille.X - Lambert_1_False_Easting
                rho_olat_MINUS_dy = Lambert_1_rho_olat - dy
                rho = Math.Sqrt(dx * dx + rho_olat_MINUS_dy * rho_olat_MINUS_dy)
                If Lambert_1_n < 0 Then
                    rho *= -1
                    dx *= -1
                    rho_olat_MINUS_dy *= -1
                End If
                If rho <> 0 Then
                    theta = Math.Atan2(dx, rho_olat_MINUS_dy) / Lambert_1_n
                    t = Lambert_1_t0 * Math.Pow(rho / Lambert_1_rho0, 1 / Lambert_1_n)
                    PHI = Pi_2 - 2 * Math.Atan(t)
                    While (Math.Abs(PHI - tempPHI) > tolerance) AndAlso (count > 0)
                        tempPHI = PHI
                        es_sin = ES_SIN_Func(Math.Sin(PHI))
                        PHI = Pi_2 - 2 * Math.Atan(t * Math.Pow((1 - es_sin) / (1 + es_sin), es_OVER_2))
                        count -= 1
                    End While
                    If count = 0 Then
                        Error_Code = Error_Code Or LAMBERT_1.NORTHING_ERROR
                        Return Error_Code
                    End If
                    PointLatLon.Lat = PHI
                    PointLatLon.Lon = theta + Lambert_1_Origin_Long
                    If Math.Abs(PointLatLon.Lat) < 0.0000002 Then
                        PointLatLon.Lat = 0
                    End If
                    If PointLatLon.Lat > Pi_2 Then
                        PointLatLon.Lat = Pi_2
                    ElseIf PointLatLon.Lat < -Pi_2 Then
                        PointLatLon.Lat = -Pi_2
                    End If
                    If PointLatLon.Lon > PI Then
                        If PointLatLon.Lon - PI < 0.0000035 Then
                            PointLatLon.Lon = PI
                        Else
                            PointLatLon.Lon -= TwoPI
                        End If
                    End If
                    If PointLatLon.Lon < -PI Then
                        If Math.Abs(PointLatLon.Lon + PI) < 0.0000035 Then
                            PointLatLon.Lon = -PI
                        Else
                            PointLatLon.Lon += TwoPI
                        End If
                    End If

                    If Math.Abs(PointLatLon.Lon) < 0.0000002 Then
                        PointLatLon.Lon = 0
                    End If
                    If PointLatLon.Lon > PI Then
                        PointLatLon.Lon = PI
                    ElseIf PointLatLon.Lon < -PI Then
                        PointLatLon.Lon = -PI
                    End If
                Else
                    If Lambert_1_n > 0 Then
                        PointLatLon.Lat = Pi_2
                    Else
                        PointLatLon.Lat = -Pi_2
                    End If
                    PointLatLon.Lon = Lambert_1_Origin_Long
                End If
            End If
            Return Error_Code
        End Function
        ''' <summary>The function Convert_Geodetic_To_Lambert_1 converts Geodetic (latitude and longitude) coordinates to Lambert Conformal Conic (1 parallel) 
        ''' projection (easting and northing) coordinates, according to the current ellipsoid and
        ''' Lambert Conformal Conic (1 parallel) projection parameters. </summary>
        ''' <param name="PointLatLon"> point à convertir en rad </param>
        ''' <param name="PointGrille"> résultat de la conversion en mètres </param>
        ''' <returns>If any errors occur, the error code(s) are returned by the function, otherwise LAMBERT_NO_ERROR is returned.</returns>
        Friend Function Convert_Geodetic_To_LAMBCC1(PointLatLon As PointD, ByRef PointGrille As PointD) As LAMBERT_1
            Dim t As Double, rho As Double, dlam As Double, theta As Double
            Dim Error_Code As LAMBERT_1
            If (PointLatLon.Lat < -Pi_2) OrElse (PointLatLon.Lat > Pi_2) Then
                Error_Code = Error_Code Or LAMBERT_1.LAT_ERROR
            End If
            If (PointLatLon.Lon < -PI) OrElse (PointLatLon.Lon > TwoPI) Then
                Error_Code = Error_Code Or LAMBERT_1.LON_ERROR
            End If
            If Error_Code = LAMBERT_1.NO_ERROR Then
                If Math.Abs(Math.Abs(PointLatLon.Lat) - Pi_2) > 0.0000000001 Then
                    t = LAMBERT_t(PointLatLon.Lat, ES_SIN_Func(Math.Sin(PointLatLon.Lat)))
                    rho = Lambert_1_rho0 * Math.Pow(t / Lambert_1_t0, Lambert_1_n)
                Else
                    If (PointLatLon.Lat * Lambert_1_n) <= 0 Then
                        Error_Code = Error_Code Or LAMBERT_1.LAT_ERROR
                        Return (Error_Code)
                    End If
                    rho = 0
                End If
                dlam = PointLatLon.Lon - Lambert_1_Origin_Long
                If dlam > PI Then
                    dlam -= TwoPI
                End If
                If dlam < -PI Then
                    dlam += TwoPI
                End If
                theta = Lambert_1_n * dlam
                PointGrille.X = rho * Math.Sin(theta) + Lambert_1_False_Easting
                PointGrille.Y = Lambert_1_rho_olat - rho * Math.Cos(theta) + Lambert_1_False_Northing
            End If
            Return Error_Code
        End Function
#End Region
    End Module
End Namespace

File: PARTAGER/ConvertirCoordonnees/ProjectionLambert_1.cs
using static FCGP.ConvertirCoordonnees;

namespace FCGP.Coordonnees
{
    /// <summary> Ce module contient les fonctions permettant la conversion entre les coordonnées Géodétique (Latitude et Longitude) et une projection 
    /// Lambert Conforme Conique à 1 parallèle en X et Y ou l'inverse traduit du projet Geotrans écrit en C </summary>
    internal static class ProjectionLambert_1
    {
        #region Variables et Constantes
        /// <summary> Donne par defaut les différentes constantes liées à la projection Lambert 1 et au système CLARKE 1880 ("NTF") </summary>
        private const double MAX_LAT = Deg_Rad * 89.9997222222222d; // ((PI * 89.9997222222222) / 180)
        private const double MIN_SCALE_FACTOR = 0.000000001d;
        private const double ONE_SECOND = 0.00000489d;
        private static double Lambert_1_a = RayonWGS84;
        private static double Lambert_1_f = F_WGS84; // 1 / 298.257223563
        private static double es = 0.0818191908426219d;
        private static double es_OVER_2 = 0.040909595421311d;
        private static double Lambert_1_n = 0.70710678118655d;
        private static double Lambert_1_rho0 = 6388838.2901212d;
        private static double Lambert_1_rho_olat = 6388838.2901211d;
        private static double Lambert_1_t0 = 0.41618115138974d;
        private static double Lambert_1_Origin_Lat = Deg_Rad * 45.0d; // (45 * PI / 180)
        private static double Lambert_1_Origin_Long = 0.0d;
        private static double Lambert_1_False_Northing = 0.0d;
        private static double Lambert_1_False_Easting = 0.0d;
        private static double Lambert_1_Scale_Factor = 1.0d;
        private const double Lambert_Delta_Easting = 400_000_000.0d;
        private const double Lambert_Delta_Northing = 400_000_000.0d;

        /// <summary> Donne différentes erreurs pouvant survenir lors de la transformation des coordonnées </summary>
        [Flags]
        internal enum LAMBERT_1 : int
        {
            /// <summary>Flag OK</summary>"
            NO_ERROR = 0,
            /// <summary>Flag erreur concernant la Latitude</summary>
            LAT_ERROR = 1,
            /// <summary>Flag erreur concernant la Longitude</summary>
            LON_ERROR = 2,
            /// <summary>Flag erreur concernant les abscisses(Easting ou X)</summary>
            EASTING_ERROR = 4,
            /// <summary>Flag erreur concernant les ordonnées(Northing ou Y)</summary>
            NORTHING_ERROR = 8,
            /// <summary>Flag erreur concernant l'origine de la Latitude</summary>
            ORIGIN_LAT_ERROR = 16,
            /// <summary>Flag erreur concernant le méridien central</summary>
            CENT_MER_ERROR = 32,
            /// <summary>Flag erreur concernant le facteur d'échelle</summary>
            SCALE_FACTOR_ERROR = 64,
            /// <summary>Flag erreur concernant le demi axe de l'éllipsoïde</summary>
            A_ERROR = 128,
            /// <summary>Flag erreur concernant l'applatissement de l'éllipsoïde</summary>
            INV_F_ERROR = 256
        }
        #endregion
        #region Functions and Procedures
        private static double LAMBERT_m(double clat, double essin)
        {
            return clat / Math.Sqrt(1.0d - essin * essin);
        }
        private static double LAMBERT_t(double lat, double essin)
        {
            return Math.Tan(Pi_4 - lat / 2.0d) / Math.Pow((1.0d - essin) / (1.0d + essin), es_OVER_2);
        }
        private static double ES_SIN_Func(double sinlat)
        {
            return es * sinlat;
        }
        /// <summary>The function Set_Lambert_1_Parameters receives the ellipsoid parameters and Lambert Conformal Conic (1 parallel)
        /// projection parameters as inputs, and sets the corresponding state variables. </summary>
        /// <param name="a">Semi-major axis of ellipsoid, in meters(input)</param>
        /// <param name="f">Flattening of ellipsoid(input)</param>
        /// <param name="Origin_Latitude">Latitude of origin, in radians(input)</param>
        /// <param name="Central_Meridian">Longitude of origin, in radians(input)</param>
        /// <param name="False_Easting">False easting, in meters(input)</param>
        /// <param name="False_Northing">False northing, in meters(input)</param>
        /// <param name="Scale_Factor">Projection scale factor(input)</param>
        /// <returns>If any errors occur, the error code(s) are returned by the function, otherwise LAMBERT_1.NO_ERROR is returned.</returns>
        internal static LAMBERT_1 Set_LAMBCC1_Parameters(double a, double f, double Origin_Latitude, double Central_Meridian, double False_Easting, double False_Northing, double Scale_Factor)
        {
            double es2;
            double es_sin;
            double m0;
            double inv_f = 1.0d / f;
            var Error_Code = default(LAMBERT_1);
            if (a <= 0d)
            {
                Error_Code |= LAMBERT_1.A_ERROR;
            }
            if (inv_f < 250.0d || inv_f > 350.0d)
            {
                Error_Code |= LAMBERT_1.INV_F_ERROR;
            }
            if (Origin_Latitude < -MAX_LAT || Origin_Latitude > MAX_LAT || Origin_Latitude > -ONE_SECOND && Origin_Latitude < ONE_SECOND)
            {
                Error_Code |= LAMBERT_1.ORIGIN_LAT_ERROR;
            }
            if (Central_Meridian < -PI || Central_Meridian > TwoPI)
            {
                Error_Code |= LAMBERT_1.CENT_MER_ERROR;
            }
            if (Scale_Factor < MIN_SCALE_FACTOR)
            {
                Error_Code |= LAMBERT_1.SCALE_FACTOR_ERROR;
            }
            if (Error_Code == LAMBERT_1.NO_ERROR)
            {
                Lambert_1_a = a;
                Lambert_1_f = f;
                Lambert_1_Origin_Lat = Origin_Latitude;
                if (Central_Meridian > PI)
                {
                    Central_Meridian -= TwoPI;
                }
                Lambert_1_Origin_Long = Central_Meridian;
                Lambert_1_False_Easting = False_Easting;
                Lambert_1_False_Northing = False_Northing;
                Lambert_1_Scale_Factor = Scale_Factor;
                es2 = 2.0d * Lambert_1_f - Lambert_1_f * Lambert_1_f;
                es = Math.Sqrt(es2);
                es_OVER_2 = es / 2.0d;
                Lambert_1_n = Math.Sin(Lambert_1_Origin_Lat);
                es_sin = ES_SIN_Func(Math.Sin(Lambert_1_Origin_Lat));
                m0 = LAMBERT_m(Math.Cos(Lambert_1_Origin_Lat), es_sin);
                Lambert_1_t0 = LAMBERT_t(Lambert_1_Origin_Lat, es_sin);
                Lambert_1_rho0 = Lambert_1_a * Lambert_1_Scale_Factor * m0 / Lambert_1_n;
                Lambert_1_rho_olat = Lambert_1_rho0;
            }
            return Error_Code;
        }
        /// <summary>The function Convert_Lambert_1_To_Geodetic converts Lambert ConformalConic (1 parallel) projection (easting and northing) coordinates to Geodetic
        /// (latitude and longitude) coordinates, according to the current ellipsoid and Lambert Conformal Conic (1 parallel) projection parameters.</summary>
        /// <param name="PointGrille"> point à convertir </param>
        /// <param name="PointLatLon"> resultat de la convertion </param>
        /// <returns>If any errors occur, the error code(s) are returned by the function PointLatLon is empty, otherwise LAMBERT_NO_ERROR is returned.</returns>
        internal static LAMBERT_1 Convert_LAMBCC1_To_Geodetic(PointD PointGrille, ref PointD PointLatLon)
        {
            double dx, dy, rho, rho_olat_MINUS_dy, t, PHI, es_sin, tempPHI = default, theta;
            double tolerance = 0.000000000485d;
            int count = 30;
            var Error_Code = LAMBERT_1.NO_ERROR;
            if (PointGrille.X < Lambert_1_False_Easting - Lambert_Delta_Easting || PointGrille.X > Lambert_1_False_Easting + Lambert_Delta_Easting)
            {
                Error_Code |= LAMBERT_1.EASTING_ERROR;
            }
            if (PointGrille.Y < Lambert_1_False_Northing - Lambert_Delta_Northing || PointGrille.Y > Lambert_1_False_Northing + Lambert_Delta_Northing)
            {
                Error_Code |= LAMBERT_1.NORTHING_ERROR;
            }
            if (Error_Code == LAMBERT_1.NO_ERROR)
            {
                dy = PointGrille.Y - Lambert_1_False_Northing;
                dx = PointGrille.X - Lambert_1_False_Easting;
                rho_olat_MINUS_dy = Lambert_1_rho_olat - dy;
                rho = Math.Sqrt(dx * dx + rho_olat_MINUS_dy * rho_olat_MINUS_dy);
                if (Lambert_1_n < 0d)
                {
                    rho *= -1;
                    dx *= -1;
                    rho_olat_MINUS_dy *= -1;
                }
                if (rho != 0d)
                {
                    theta = Math.Atan2(dx, rho_olat_MINUS_dy) / Lambert_1_n;
                    t = Lambert_1_t0 * Math.Pow(rho / Lambert_1_rho0, 1d / Lambert_1_n);
                    PHI = Pi_2 - 2d * Math.Atan(t);
                    while (Math.Abs(PHI - tempPHI) > tolerance && count > 0)
                    {
                        tempPHI = PHI;
                        es_sin = ES_SIN_Func(Math.Sin(PHI));
                        PHI = Pi_2 - 2d * Math.Atan(t * Math.Pow((1d - es_sin) / (1d + es_sin), es_OVER_2));
                        count -= 1;
                    }
                    if (count == 0)
                    {
                        Error_Code |= LAMBERT_1.NORTHING_ERROR;
                        return Error_Code;
                    }
                    PointLatLon.Lat = PHI;
                    PointLatLon.Lon = theta + Lambert_1_Origin_Long;
                    if (Math.Abs(PointLatLon.Lat) < 0.0000002d)
                    {
                        PointLatLon.Lat = 0d;
                    }
                    if (PointLatLon.Lat > Pi_2)
                    {
                        PointLatLon.Lat = Pi_2;
                    }
                    else if (PointLatLon.Lat < -Pi_2)
                    {
                        PointLatLon.Lat = -Pi_2;
                    }
                    if (PointLatLon.Lon > PI)
                    {
                        if (PointLatLon.Lon - PI < 0.0000035d)
                        {
                            PointLatLon.Lon = PI;
                        }
                        else
                        {
                            PointLatLon.Lon -= TwoPI;
                        }
                    }
                    if (PointLatLon.Lon < -PI)
                    {
                        if (Math.Abs(PointLatLon.Lon + PI) < 0.0000035d)
                        {
                            PointLatLon.Lon = -PI;
                        }
                        else
                        {
                            PointLatLon.Lon += TwoPI;
                        }
                    }

                    if (Math.Abs(PointLatLon.Lon) < 0.0000002d)
                    {
                        PointLatLon.Lon = 0d;
                    }
                    if (PointLatLon.Lon > PI)
                    {
                        PointLatLon.Lon = PI;
                    }
                    else if (PointLatLon.Lon < -PI)
                    {
                        PointLatLon.Lon = -PI;
                    }
                }
                else
                {
                    if (Lambert_1_n > 0d)
                    {
                        PointLatLon.Lat = Pi_2;
                    }
                    else
                    {
                        PointLatLon.Lat = -Pi_2;
                    }
                    PointLatLon.Lon = Lambert_1_Origin_Long;
                }
            }
            return Error_Code;
        }
        /// <summary>The function Convert_Geodetic_To_Lambert_1 converts Geodetic (latitude and longitude) coordinates to Lambert Conformal Conic (1 parallel) 
        /// projection (easting and northing) coordinates, according to the current ellipsoid and
        /// Lambert Conformal Conic (1 parallel) projection parameters. </summary>
        /// <param name="PointLatLon"> point à convertir en rad </param>
        /// <param name="PointGrille"> résultat de la conversion en mètres </param>
        /// <returns>If any errors occur, the error code(s) are returned by the function, otherwise LAMBERT_NO_ERROR is returned.</returns>
        internal static LAMBERT_1 Convert_Geodetic_To_LAMBCC1(PointD PointLatLon, ref PointD PointGrille)
        {
            double t;
            double rho;
            double dlam;
            double theta;
            var Error_Code = default(LAMBERT_1);
            if (PointLatLon.Lat < -Pi_2 || PointLatLon.Lat > Pi_2)
            {
                Error_Code |= LAMBERT_1.LAT_ERROR;
            }
            if (PointLatLon.Lon < -PI || PointLatLon.Lon > TwoPI)
            {
                Error_Code |= LAMBERT_1.LON_ERROR;
            }
            if (Error_Code == LAMBERT_1.NO_ERROR)
            {
                if (Math.Abs(Math.Abs(PointLatLon.Lat) - Pi_2) > 0.0000000001d)
                {
                    t = LAMBERT_t(PointLatLon.Lat, ES_SIN_Func(Math.Sin(PointLatLon.Lat)));
                    rho = Lambert_1_rho0 * Math.Pow(t / Lambert_1_t0, Lambert_1_n);
                }
                else
                {
                    if (PointLatLon.Lat * Lambert_1_n <= 0d)
                    {
                        Error_Code |= LAMBERT_1.LAT_ERROR;
                        return Error_Code;
                    }
                    rho = 0d;
                }
                dlam = PointLatLon.Lon - Lambert_1_Origin_Long;
                if (dlam > PI)
                {
                    dlam -= TwoPI;
                }
                if (dlam < -PI)
                {
                    dlam += TwoPI;
                }
                theta = Lambert_1_n * dlam;
                PointGrille.X = rho * Math.Sin(theta) + Lambert_1_False_Easting;
                PointGrille.Y = Lambert_1_rho_olat - rho * Math.Cos(theta) + Lambert_1_False_Northing;
            }
            return Error_Code;
        }
        #endregion
    }
}